### PR TITLE
Replace Monkeypatching of BasicObject with TracePoint

### PR DESCRIPTION
**Pull request by muzfuz**
Here's a first stab at this work.

I have broken a LOT of tests, which need fixing.

That said -- query generation definitely works! 🎉

### CODE REVIEW

**Comment by muzfuz** [comment]
Here's a first stab at this work.

I have broken a LOT of tests, which need fixing.

That said -- query generation definitely works! 🎉

File: lib/delfos/method_trace.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+require "binding_of_caller"
+require "delfos/method_logging"
+require "delfos/method_logging/code_location"
+
+class Pathname
+  def is_inside?(path)
```
Comment (markburns) [inline lib/delfos/method_trace.rb]: I'd not do this with a monkey patch but still if this works it's an impressive minimisation of code.

File: lib/delfos/method_trace.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+module Delfos
+  class MethodTrace
+    class << self
+
+      def trace(app_dirs)
+
+        TracePoint.new(:call) do |tp|
+          app_dirs.each do |dir|
+            if Pathname(tp.path).is_inside?(dir)
+              show_caller = ->(s, label=nil) {
```
Comment (markburns) [inline lib/delfos/method_trace.rb]: Rename this and remove second arg

File: lib/delfos/method_trace.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                method_name: tp.method_id,
+                file: tp.path,
+                line_number: tp.lineno,
+              )
+
+              ::Delfos::MethodLogging.log(call_site, called_code) if show_caller.("__method__")
+            end
+          end
+        end.enable
+
```
Comment (markburns) [inline lib/delfos/method_trace.rb]: Will probably need some extract methods to satisfy code climate

File: lib/delfos/method_trace.rb (hunk 1 above)
Comment (muzfuz) [inline lib/delfos/method_trace.rb]: Ah yes, really I just forgot to remove this.  Will definitely rework 👍

File: lib/delfos/method_trace.rb (hunk 2 above)
Comment (muzfuz) [inline lib/delfos/method_trace.rb]: 👍

**Comment by markburns** [comment]
So I tried to get this back to green.
A few things to note. I've added TODOs where there are definitely things to fix.

I've tried to get the `CallStack` re-implemented using `TracePoint` `return` and `raise`.

I like a lot of the removal of complexity. Unfortunately some of the complexity is necessary. 
I didn't manage to get fully explanatory specs around exactly what some of this stuff has been added for. And some of it is just stabbing-in-the-dark until we get bundler and rails to green.

Anyway. You'd removed the explicit setting of the `class_method` boolean. I don't think we can get around this. We have to determine whether it is a class method based on who `self` is, and not based on whether or not that method is present as an `instance_method`.

I think that rather than have two things: a `call_site` and a `called_code` concept, we'll need more.

We have to record the
  * Container method containing the call_site and it's source location (file and line number), class and whether or not it is a class method.
  * The call_site information - file and line number
  * Method being called at the `call_site`, source location, class, and whether or not it is a class method.

I'd shovelled in the containing method information into the call_site information and had to hack on the `method_definition_line` stuff etc. 
So I think you are correct to remove it from there.
However, we still want the info.
I think it will be easier to just deal with the three things, than forcing call_site to also record information about its container method.

Hope this makes sense.

File: lib/delfos/method_logging/code_location.rb [hunk 4]
```diff
@@ -1,47 +1,32 @@
...
-      def method_definition_line
-        method_definition&.last&.to_i || fallback_method_definition_line_number
+      def method_name
+        @method_name.to_s
       end
 
       def method_type
-        class_method ? "ClassMethod" : "InstanceMethod"
+        #TODO write a spec for and fix this
+        # This doesn't work if we have SomeClass.some_method and SomeClass#some_method
```
Comment (markburns) [inline lib/delfos/method_logging/code_location.rb]: I think we will have to go back to what we had before

File: lib/delfos/method_logging/code_location.rb [hunk 5]
```diff
@@ -67,18 +52,6 @@ def relative_path(file, dir)
...
-
-      def fallback_method_definition_file
-        @file
-      end
-
-      def fallback_method_definition_line_number
-        0
-      end
-
-      def method_definition
```
Comment (markburns) [inline lib/delfos/method_logging/code_location.rb]: If we need to keep using meta-programming to fetch the container method source locations, we might want to bring back the MethodCache.
It was previously used for source_location lookups but also for calling the original implementation.
We don't need it for calling the original any more, but may need it for performance reasons.

File: lib/delfos/method_trace.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      STACK_OFFSET = 6
+
+      def trace
+        on_raise.enable
+        on_call.enable
+        on_return.enable
+      end
+
+      def disable!
+        @on_call&.disable
```
Comment (markburns) [inline lib/delfos/method_trace.rb]: we can almost certainly make this stuff cleaner

File: lib/delfos/method_trace.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+      def handle_call(tp)
+        return unless relevant?(tp)
+
+        call_site   = code_location_from_call_site
+        CallStack.push(call_site)
+
+        called_code = code_location_from_trace_point(tp)
+
+        ::Delfos.call_site_logger.log(call_site, called_code) if call_site.method_name
```
Comment (markburns) [inline lib/delfos/method_trace.rb]: This API should change to:

```ruby
::Delfos.call_site_logger.log(container_method, call_site, called_code)
```

File: lib/delfos/neo4j/call_stack_query.rb [hunk 8]
```diff
@@ -48,7 +48,7 @@ def call_site_query(_cs, i)
             m#{i} :Method {
               type: {method_type#{i}},
               name: {method_name#{i}},
-              file: {method_definition_file#{i}},
+              file: {file#{i}},
               line_number: {method_definition_line#{i}}}
```
Comment (markburns) [inline lib/delfos/neo4j/call_stack_query.rb]: This stuff will be easier to reason about once we eliminate `method_definition_line` and `method_definition_file` from the `call_site` and introduce a 3rd `container_method` param to `CallSiteLogger.log`

**Comment by markburns** [comment]
I suspect that we might also not be covering this case correctly:

```ruby
#a.rb
class A
  def super_class_method #called_code
     C.something_else   
  end
end

#b.rb
class B < A
end

#c.rb
class C
  def self.something_else 
  end
end

#bin/script
B.new.super_class_method #call site
```
container_method should look like 
```ruby
{object: "(main)", method_name: "(main)", line_number:0, file: "bin/script", class_method: false}
```

call_site should be 
```ruby
{object: B, method_name: "super_class_method", line_number: 1, file: "bin/script"}
```

called_code should be 
```ruby
{object: B, method_name: "super_class_method", class_method: false, line_number: 2, file: "a.rb"}
```

Also I think it is clear that `call_site` is a different type of `CodeLocation` to `container_method` and `called_code`

**Comment by markburns** [comment]
This is a nice to have but perhaps ideally `call_site` would include the character where the method is called.

`product.name || Lookup.default_name`
```
{line_number: 15, file: "search.rb", character: 8}

{line_number: 15, file: "search.rb", character: 22}

```

**Comment by markburns** [comment]
@muzfuz - I pushed quite a few changes. It's not quite green yet. Just integration specs left.
I'm not sure what the issue is yet. 
Looks like the `CallStack.pop` gets called too many times during the `return` `TracePoint`. 

Perhaps it is too permissive in comparison to the `call` `TracePoint` so manages to execute more times than the `CallStack.push`

File: lib/delfos/method_trace.rb [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      def on_return
+        @on_return ||= TracePoint.new(:return) do |tp|
+          handle_return tp
+        end
+      end
+
+      def handle_return(tp)
+        call_site = call_site_from(tp)
+        return unless relevant?(call_site)
+        CallStack.pop
```
Comment (markburns) [inline lib/delfos/method_trace.rb]: For some reason this gets triggered too many times. So maybe somehow the value of `relevant?(tp)` is true more often with the `return` `TracePoint`s

**Comment by muzfuz** [comment]
Thanks for all these changes @markburns -- have gotten sidetracked by other issues this morning, but will try to get back on later this afternoon.

**Comment by markburns** [comment]
No worries. And thanks for working on it too!
I think it's close to being fixed. I feel like it should be something simple.
We should maybe add a few more unit specs to get to the bottom of it. I've done a lot of moving things around and things are probably being tested at the wrong level now.
Particularly around the `MethodTrace` class

**Comment by chrisface** [comment]
I'll see If I can give a hand on this one. I started building a very similar tool using TracePoint as well just a week ago. Seems you're much further along with me so will start contributing to here 👍

**Comment by muzfuz** [comment]
Looks like the latest changes have caused my local test app to stop logging out events.  Taking a look at why that is.

**Comment by muzfuz** [comment]
Happy to have more input @chrisface !

**Comment by markburns** [comment]
Hey guys. Sorry it's late here. I may have left it in a messier state than it was before with the TracePoint stuff.
If you need to roll back a bit that's fine. Hopefully it's reasonably obvious what I was going for.
It's quite hard to debug because RSpec just exits if there's anything goes wrong inside the TracePoint code. And byebug doesn't work either.

**Comment by muzfuz** [comment]
Yeah I find a combination of running specs and running a small test ruby app works quite well.

I think the code is overall in a better state, and I suspect the reason for things not getting logged out it quite straightforward.  Will hopefully solve soon.

**Comment by muzfuz** [comment]
Right, I commented out `on_return.enable` on this line here, and method logging has started working again.

https://github.com/ruby-analysis/delfos/blob/use_tracepoint/lib/delfos/method_trace.rb#L15

Will try to figure out why now.

File: lib/delfos/method_trace.rb [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      end
+
+      def on_call
+        @on_call ||= TracePoint.new(:call) do |tp|
+          handle_call(tp)
+        end
+      end
+
+      def temporarily_disable_trace_point
+        return if @disable_trace_point
```
Comment (markburns) [inline lib/delfos/method_trace.rb]: NB I started spiking with `tp.disable` and `tp.enable` instead of flagging with these instance variables.
Still not got to green locally either. But either way, I don't think we should do this

File: lib/delfos/method_trace.rb [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+          file:         eval_in_caller("method(__method__).source_location.first", 1),
+          line_number:  eval_in_caller("method(__method__).source_location.last", 1),
+          class_method: eval_in_caller('is_a?(Module)', 1),
+        )
+      end
+
+      def eval_in_caller(s, extra_offset=0)
+        return if s.nil?
+
+        other = binding.of_caller(STACK_OFFSET + extra_offset)
```
Comment (markburns) [inline lib/delfos/method_trace.rb]: This offset may need fixing. 
I'm thinking a slightly more sensible way to do this class is to unit test it and break apart into unit tested classes.
So there's a minimal surface area between the hard to debug/test/dev TracePoint code and the actual stuff we want to do.

**Comment by markburns** [comment]
@bestie - we just started hacking on the `TracePoint` idea. 

Did you get anywhere with it yourself?

File: lib/delfos/method_trace.rb (hunk 11 above)
Comment (markburns) [inline lib/delfos/method_trace.rb]: Although having said that `binding.of_caller` is tricky to work with because it needs values hard-coded based on the specific implementation in place.

File: lib/delfos/method_trace.rb (hunk 10 above)
Comment (muzfuz) [inline lib/delfos/method_trace.rb]: Agreed, if there's a cleaner way we can achieve this then we should use that.

**Comment by markburns** [comment]
@muzfuz - yeah that makes sense that it started working again. 
The only issue is that it breaks `CallStack` recording. 
I.e. the chains of execution.

**Comment by muzfuz** [comment]
Indeed, seeing that now.  Hm...

**Comment by muzfuz** [comment]
Hey @markburns could you please clarify why you want to temporarily disable TracePoint?  I'm actually not sure I understand why we need to flip it off / on

**Comment by muzfuz** [comment]
I think here's the approach I'll take on this.

```
trace.enabled?
#=> true

trace.disable do
    trace.enabled?
    # only disabled for this block
end

trace.enabled?
#=> true
```

**Comment by markburns** [comment]
Hey @muzfuz -
I think we don't need it. 
Sorry that's a red herring. In writing my explanation, I wrote this snippet to show the thinking. However the output shows that the individual TracePoint *is* disabled inside TracePoint blocks.

E.g.

```ruby
def outer
  puts "outer: before inner"
  inner
  puts "outer: after inner"
end

def inner
  puts "inner"
end

def internal_trace_point_method
  puts "TP: don't trigger trace point in here"
  another_tp_method
end

def another_tp_method
  puts "TP: don't trigger trace point in here"
end

TracePoint.new(:call) do |tp|
  puts "TP: call #{tp.method_id}"
  internal_trace_point_method
end.enable

TracePoint.new(:return) do |tp|
  puts "TP: return from #{tp.method_id}"
end.enable

outer
```

We see

```
TP: call outer
TP: don't trigger trace point in here
TP: don't trigger trace point in here
outer: before inner
TP: call inner
TP: don't trigger trace point in here
TP: don't trigger trace point in here
inner
TP: return from inner
outer: after inner
TP: return from outer
```

File: lib/delfos/method_trace.rb [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+          CallStack.pop rescue nil
+        end
+      end
+
+      private
+
+      def relevant?(call_site)
+        AppDirectories.include_files?(
+          call_site.path,
```
Comment (muzfuz) [inline lib/delfos/method_trace.rb]: Unfortunately I don't think we can use this, as it will actually go off and use `binding.of_caller` to evaluate this stuff -- and that in return will break if you're running something like a `Pry` session.  Trying to think of other ways to evaluate this stuff.

Comment (markburns) [inline lib/delfos/method_trace.rb]: Are you sure it's not the case that it will actually work but just not during debugging?

Comment (muzfuz) [inline lib/delfos/method_trace.rb]: Right so looking closely at this - I think we do need this, but before we reach this step we will want to do some blanket filtering based on tracepoint file path.

We _will_ want to go and do this stuff for our `call_site`, but before we get to that point we should also check that we are in the right file by using `tp.path` -- otherwise we get bindings into all sorts of places like `main` and `Irb` or `Pry`.

Comment (markburns) [inline lib/delfos/method_trace.rb]: By the way do you mean 'break' as in value will differ? The offset is different while using a debugger compared to a sing 'puts'

**Comment by markburns** [comment]
I think it's possible you're correct but any assumptions you're making
based on using pry etc may be way off.
Using pry to look at this may give you a distorted view

**Comment by muzfuz** [comment]
Not using pry :) I'm running specs and simultaneously executing this small program on the side, which is helping to bubble up these issues.

In this case, executing `DelfosTest.exec` in the context of `main` breaks (it so happens that I have Pry set up as my console, but this also happens in Irb).

```
require "logger"
logger = Logger.new(STDOUT)
require "delfos"
Delfos.setup!(logger: logger)

class Product
  def initialize(price = nil, name = nil)
    @price = price
    @name = name
  end

  def change_name(name)
    @name = name
  end
end

class DelfosTest
  class << self
    def exec
      product = Product.new
      product.change_name("Car")
    end
  end
end

DelfosTest.exec
```

**Comment by markburns** [comment]
Nice.

I think your logger setup should be unnecessary and if not we should work
to make this default.

Also let's make a simple program like this a default that is available in
the examples directory.

Based on how simple your program is and how little time I spent on trying
to fix the build yesterday I think you'd more productively spend your time
fixing stuff

I've just spent a portion of my day trying to improve things but may have
made stuff worse.

Sorry if this feedback is not super useful!

File: lib/delfos/method_trace.rb (hunk 12 above)
Comment (muzfuz) [inline lib/delfos/method_trace.rb]: It breaks because `method` don't be defined.  It makes sense if you run the small code sample I pasted below, and is quite easy to mitigate, just trying to find the cleanest way to do it.

**Comment by muzfuz** [comment]
I found the logger step necessary, but yes I agree we should make it default.

I agree it might be a good idea to include this program as a test step -- I'll look at doing that soon.

I've also been making changes this end which I _think_ has made everything work correctly.  Just running a few more tests and then will push it up -- happy to do another branch if it'll help avoid confusion.

File: lib/delfos/method_trace.rb (hunk 11 above)
Comment (muzfuz) [inline lib/delfos/method_trace.rb]: Yeah the stack offset really kind of does my head in a little bit.  In a way it good to be so explicit about it, because it forces you to think about how it all works.

**Comment by muzfuz** [comment]
Right so it's a fairly small refactor, but I think it's cleaned things up a bit, and it looks like events are now firing correctly.

My next step is to look at the red specs, add a few integration cases, and also look at handling the `raise` events.

**Comment by markburns** [comment]
@muzfuz - this is back to green now. Before merging I want to tidy up the history and add in some specs for the MethodTrace stuff. Also break apart into smaller classes. 
So the code that sets up and disables `TracePoint` will be separated from the code that builds the container_method, call_site and called_method.
Calling it a day here on this. So if you happen to be looking at it this weekend, then feel free to give it a shot.
In fact I'm less bothered about the breaking into smaller classes than tidying up the history and adding specs. So would merge with just those changes

**Comment by muzfuz** [comment]
Liking these changes @markburns 

I'll get started on cleaning up the commit history and then think about writing a spec for `MethodTrace`.  

I agree it may be worth merging now and then iterating, to avoid having this pull request balloon into an even bigger set of changes.

**Comment by bestie** [comment]
@markburns unfortunately not, I was thinking about it but didn't have chance to dig in. This sounds great though, looking forward to making use of it!

**Comment by muzfuz** [comment]
@markburns Right I have cleaned up the commit history.  If you disagree with these then feel free to overwrite with your local branch.

I'll now look to add a unit test for `MethodTrace`.  Need to think about testing strategy on this one, as obviously we don't really want to test `TracePoint` itself.

**Comment by markburns** [comment]
Yeah I'd just test that something gets called when you enable TracePoint
then beyond that method call test in an isolated class or classes
completely decoupled from TracePoint. The only difficulty might be the
stack offset which we will probably need to make overridable

**Comment by markburns** [comment]
I've done the decoupling work and all the unit tests are passing. Although integration specs aren't.
I want to replicate the higher level failures with lower level failing unit tests and hopefully fix both sets in one.
Probably calling it a day here now. But will probably have a look in the morning tomorrow too.

**Comment by markburns** [comment]
If you get to look at this before me. A tip (apologies if it's obvious), is to debug during a spec before a failure, then paste the query into to the neo4j `http://localhost:7476/browser` query editor.
Then delete sections of the query until you find which part isn't present.
It seems that currently the call sites call their own container methods:
`/spec/integration/neo4j/call_sites_spec.rb:48`

image


```cypher
 MATCH (b:Class{name: "B"})  -  [:OWNS]
          ->  (mb:Method{type: "InstanceMethod", name: "another_method"})

          -[:CONTAINS]->(csB2C:CallSite)-[:CALLS]->(mc)
        MATCH (a:Class{name: "A"})  -  [:OWNS]
          -> (ma:Method{name: "some_method"}) -[:CONTAINS]->cs
       
        RETURN a,ma,mb,mc, cs
```

**Comment by markburns** [comment]
@muzfuz - still not much closer to fixing the integration specs.
Unfortunately since using TracePoint this library has gotten much harder to debug/work with. As pry or byebug no longer work.

The remaining failures are possibly just to do with the stack offset values. 

I'd like to get a good way of testing them, but `binding_of_caller` only works on the current binding. E.g. you can't pass a binding around and get its first caller. 
You just get the first caller of the current call site.
E.g.

```ruby
require "binding_of_caller"

def i_want_this_binding
  message = "should be this"
  called_method()
end

def called_method
  binding
end

def but_i_get_this
  message = "unexpected"
  other_call_site
end

def other_call_site
  puts i_want_this_binding.of_caller(1).eval "message"
end

but_i_get_this
```

Expect `"should be this"` but get `"unexpected"`

**Comment by muzfuz** [comment]
@markburns we are taking a look this morning, going to try to pair up with someone this end to get it over the line.

**Comment by muzfuz** [comment]
Right so I think I know what's going on with `custom_logger_spec`

We are setting up with `application_directories: ["fixtures"]`, but then we actually execute in the context of the spec file by using `loading_code`.  

This means that we don't actually do any method tracing for our execution context.  That's the first problem.

If move you the `loading_code` to `a.rb` then the test passes, BUT only if you execute `A.new.some_method`.  If you subsequently also execute `B.new.another_method anything` then it thinks that `call_sites.length` is 7.

This is because we in fact have multiple `call_sites` arrays.  So interesting behaviour that I wouldn't have expected, but it actually seems to make sense.

You can see this behaviour in the commit I just pushed up.

Try running this spec and then uncommenting line 27 in `a.rb`.

**Comment by muzfuz** [comment]
Right I've actually added a test runner and that's sorted it out.  Execution context was obviously the issue there.  Take a look at `7059ecd` to see the thinking from earlier.

**Comment by muzfuz** [comment]
Looking at the rest of these tests -- we might be better off reworking these to run against the actual test code, and then ensuring that we are sending the correct objects to our logger.

**Comment by markburns** [comment]
Sure. These tests were mostly added for edge cases and when used with the
tons of meta programming approach.
It probably does make sense to make the boundary point the neo4j logger and
also add more unit tests for the query generation at the other side of that
boundary point.

**Comment by muzfuz** [comment]
💯
